### ISSUE TITLE
FastCorrect2: Poor performance in pre-training phase

### ISSUE THREAD

**Issue by Arrivederci** [microsoft/NeuralSpeech#16]
I pretrained FastCorrect2 model with the settings in "runs/train_pretrain.sh". The pseudo paired dataset was generated using the corpus downloaded at [here](https://github.com/brightmart/nlp_chinese_corpus). I think the dataset size is enough for pretraining. But the result looks compromised, loss stopped dropping at the 7th epoch. And the model also performs very bad after fine-tuning on Aishell-1 dataset. Here is my log at validation step. 
```
2022-03-28 08:15:55 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.448 | nll_loss 0.596 | wer_dur_loss 0.014 | word_ins 1.855 | word_ins1 1.855 | word_ins2 1.855 | word_ins3 1.855 | to_be_edited_loss 0.037 | closest_loss 0.008 | ppl 174.59 | wps 312994 | wpb 39214.3 | bsz 1042.1 | num_updates 11156
2022-03-28 19:06:00 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.418 | nll_loss 0.602 | wer_dur_loss 0.01 | word_ins 1.844 | word_ins1 1.844 | word_ins2 1.844 | word_ins3 1.844 | to_be_edited_loss 0.028 | closest_loss 0.045 | ppl 171.01 | wps 308015 | wpb 39214.3 | bsz 1042.1 | num_updates 22305 | best_loss 7.418
2022-03-29 05:49:45 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.406 | nll_loss 0.598 | wer_dur_loss 0.009 | word_ins 1.841 | word_ins1 1.841 | word_ins2 1.841 | word_ins3 1.841 | to_be_edited_loss 0.026 | closest_loss 0.049 | ppl 169.61 | wps 315539 | wpb 39214.3 | bsz 1042.1 | num_updates 33455 | best_loss 7.406
2022-03-29 16:33:39 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 7.391 | nll_loss 0.6 | wer_dur_loss 0.009 | word_ins 1.839 | word_ins1 1.839 | word_ins2 1.839 | word_ins3 1.839 | to_be_edited_loss 0.025 | closest_loss 0.037 | ppl 167.9 | wps 311628 | wpb 39214.3 | bsz 1042.1 | num_updates 44603 | best_loss 7.391
2022-03-30 03:08:25 | INFO | valid | epoch 005 | valid on 'valid' subset | loss 7.39 | nll_loss 0.609 | wer_dur_loss 0.009 | word_ins 1.838 | word_ins1 1.838 | word_ins2 1.838 | word_ins3 1.839 | to_be_edited_loss 0.025 | closest_loss 0.04 | ppl 167.72 | wps 321428 | wpb 39214.3 | bsz 1042.1 | num_updates 55753 | best_loss 7.39
2022-03-30 13:30:34 | INFO | valid | epoch 006 | valid on 'valid' subset | loss 7.378 | nll_loss 0.611 | wer_dur_loss 0.009 | word_ins 1.837 | word_ins1 1.837 | word_ins2 1.837 | word_ins3 1.837 | to_be_edited_loss 0.025 | closest_loss 0.026 | ppl 166.39 | wps 320994 | wpb 39214.3 | bsz 1042.1 | num_updates 66904 | best_loss 7.378
2022-03-31 00:00:03 | INFO | valid | epoch 007 | valid on 'valid' subset | loss 7.375 | nll_loss 0.614 | wer_dur_loss 0.008 | word_ins 1.837 | word_ins1 1.837 | word_ins2 1.837 | word_ins3 1.837 | to_be_edited_loss 0.023 | closest_loss 0.027 | ppl 165.98 | wps 320013 | wpb 39214.3 | bsz 1042.1 | num_updates 78051 | best_loss 7.375
2022-03-31 10:21:02 | INFO | valid | epoch 008 | valid on 'valid' subset | loss 7.378 | nll_loss 0.609 | wer_dur_loss 0.009 | word_ins 1.837 | word_ins1 1.837 | word_ins2 1.837 | word_ins3 1.837 | to_be_edited_loss 0.025 | closest_loss 0.031 | ppl 166.37 | wps 321458 | wpb 39214.3 | bsz 1042.1 | num_updates 89201 | best_loss 7.375
2022-03-31 20:32:51 | INFO | valid | epoch 009 | valid on 'valid' subset | loss 7.384 | nll_loss 0.611 | wer_dur_loss 0.009 | word_ins 1.836 | word_ins1 1.836 | word_ins2 1.836 | word_ins3 1.836 | to_be_edited_loss 0.024 | closest_loss 0.046 | ppl 167.01 | wps 325090 | wpb 39214.3 | bsz 1042.1 | num_updates 100350 | best_loss 7.375
```
I pretrained model on 8 GPUs, with the settings in "runs/train_pretrain.sh", the batch size is 8000+, maybe it's too large? Could you please give me some advice？

**Comment by YichongLeng** [comment]
For batch size, we use 5000 tokens * 8 GPU * 8 num_updates in pretraining and 5000 tokens * 4 GPU * 8 num_updates in finetuning.

Although I never use the dataset you mention, after a brief look I think it is okay for FastCorrect2 pretraining. A quick way to check the quality of the dataset is to use that dataset to pretrain (and finetune) a standard autoregressive correction model. If you can use this dataset to train an autoregressive model, then it will be suitable for FastCorrect 2. Otherwise the dataset may not be well cleaned.

For standard autoregressive correction model, it can be easily trained since it is just a translation model and supported by FairSeq.

**Comment by Amber921463001** [comment]
我已经按照您设定的参数微调了模型，但我预测的结果和原句子一样，请问您知道这是什么原因吗？
我预测的输出是translated = [transf_gec.decode(hypos[0]['tokens']) for hypos in batched_hypos][0]
